federation/pkg/federation-controller/namespace/namespace_controller.go

```diff
@@ -0,0 +1,353 @@
+// error (like failure to create).
+type namespaceItem struct {
+	namespace string
+	trial     int64
+}
+
+// NewNamespaceController returns a new namespace controller
+func NewNamespaceController(client federation_release_1_4.Interface) *NamespaceController {
+
+	// Builds stop chanels.
+	nc := &NamespaceController{
+		namespaceStopChan:          make(chan struct{}),
+		clusterReconcileStopChan:   make(chan struct{}),
+		namespaceReconcileStopChan: make(chan struct{}),
+	}
+
+	// Build delivereres for triggering reconcilations.
+	nc.namespaceDeliverer = util.NewDelayingDeliverer(make(chan *util.DelayingDelivererItem, namespaceChanSize))
+	nc.clusterDeliverer = util.NewDelayingDeliverer(make(chan *util.DelayingDelivererItem, clusterChanSize))
+
+	// Start informer in federated API servers on namespaces that should be federated.
+	nc.namespaceInformerStore, nc.namespaceInformerController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (pkg_runtime.Object, error) {
+				return client.Core().Namespaces().List(options)
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return client.Core().Namespaces().Watch(options)
+			},
+		},
+		&api.Namespace{},
+		controller.NoResyncPeriodFunc(),
+		framework.ResourceEventHandlerFuncs{
+			// Trigger reconcilation whenever some defiition is changed.
+			DeleteFunc: func(old interface{}) { nc.deliverNamespaceObj(old, 0, 0) },
+			AddFunc:    func(cur interface{}) { nc.deliverNamespaceObj(cur, 0, 0) },
+			UpdateFunc: func(old, cur interface{}) {
+				nold, _ := old.(*api_v1.Namespace)
+				ncur, _ := old.(*api_v1.Namespace)
+				// Object meta or spec updated. Status changes are ignored.
+				if !reflect.DeepEqual(nold.Spec, ncur.Spec) ||
+					!reflect.DeepEqual(nold.ObjectMeta, ncur.ObjectMeta) {
+					nc.deliverNamespaceObj(cur, 0, 0)
+				}
+			},
+		},
+	)
+
+	// Federated informer on namespaces in members of federation.
+	nc.namespaceFederatedInformer = util.NewFederatedInformer(
+		client,
+		func(cluster *federation_api.Cluster, targetClient federation_release_1_4.Interface) (cache.Store, framework.ControllerInterface) {
+			return framework.NewInformer(
+				&cache.ListWatch{
+					ListFunc: func(options api.ListOptions) (pkg_runtime.Object, error) {
+						return targetClient.Core().Namespaces().List(options)
+					},
+					WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+						return targetClient.Core().Namespaces().Watch(options)
+					},
+				},
+				&api.Namespace{},
+				controller.NoResyncPeriodFunc(),
+				framework.ResourceEventHandlerFuncs{
+					// Trigger reconcilation whenever something in federated cluster is changed. In most cases it
+					// would be just confirmation that some namespace opration suceeded.
+					DeleteFunc: func(old interface{}) {
+						// Remove once it is done automatically.
+						nold, _ := old.(*api_v1.Namespace)
```

Comment:
If you are doing:
```
nold, _ := old.(*api_v1.Namespace)
```
you are explicitly ignoring error here. This is bad. You should either:
- check error that it returns
- or just do:
```
nold := old.(*api_v1.Namespace)
```
that will result in panic if the old object is not namespace.

What you are doing is just incorrect, because it can lead to unexpected panics (if casting returns errors, then nold would be nil).

The same everywhere else in this file.